### PR TITLE
Reactive xc 39

### PR DESCRIPTION
**Pull request by mbarekh**
replace prototype by classes

### CODE REVIEW

File: test/spec/xcWebSocketSubscriberSpec.ts [hunk 1]
```diff
@@ -12,11 +12,14 @@ describe("Test xcWebSocketSubscriber module", function () {
             var serverUrl = "wss://" + (new Guid()).create();
             mockServer = Mock.createMockServer(serverUrl);
             mockWebSocket = Mock.createMockWebSocket(serverUrl);
+            console.log("AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA");
```
Comment (maurelio1234) [inline test/spec/xcWebSocketSubscriberSpec.ts]: no debugging logs on a commit please.

File: test/spec/xcSessionSpec.ts [hunk 2]
```diff
@@ -37,13 +42,14 @@ describe("Test xcSession module", function () {
         let session;
         beforeEach(function () {
             let serverUrl = "wss:\\serverUrl";
-            let mockWebSocket = new WebSocket(serverUrl);
+            window["WebSocket"] = WebSocket;
```
Comment (maurelio1234) [inline test/spec/xcSessionSpec.ts]: @nicolaserny  is doing this differently in the other PR...

https://github.com/xcomponent/ReactiveXComponent.js/pull/31/files#diff-a9ae241995226095ebf400b83a19e1bdR16

we'd better synchronize sooner than later.

File: src/communication/xcWebSocketSubscriber.ts [hunk 3]
```diff
@@ -21,7 +21,7 @@ class Subscriber {
         this.subscribedStateMachines = {};
         this.observableMsg = Rx
             .Observable
-            .fromEvent(this.webSocket, "message");
+            .fromEvent(this.webSocket.getWS(), "message");
```
Comment (maurelio1234) [inline src/communication/xcWebSocketSubscriber.ts]: could we avoid unnecessary abbreviations? renaming is cheap.

File: test/spec/mock/mockSubscriberDependencies.ts [hunk 4]
```diff
@@ -2,6 +2,9 @@ import { WebSocket, Server, SocketIO } from "mock-socket";
 import Subscriber from "communication/xcWebSocketSubscriber";
 import webSocketConf from "configuration/xcWebSocketBridgeConfiguration";
 
+import IWS from "communication/IWebSocket";
```
Comment (maurelio1234) [inline test/spec/mock/mockSubscriberDependencies.ts]: could we avoid unnecessary abbreviations? renaming is cheap.

File: src/communication/IWebSocket.ts [hunk 5]
```diff
@@ -0,0 +1,10 @@
+interface IWebSocket
+{
+
+    send(message : string) : void;
+    close() : void;
+    setEventListener(event : string, eventListener : (e) => void) : void;
+    getWS() : any;
```
Comment (maurelio1234) [inline src/communication/IWebSocket.ts]: could we avoid unnecessary abbreviations? renaming is cheap.

File: src/communication/WebSocket.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
+import IWebSocket from "./IWebSocket";
+import {javascriptHelper} from "javascriptHelper";
+
+class WebSocket implements IWebSocket {
+
+    private ws : any;
```
Comment (maurelio1234) [inline src/communication/WebSocket.ts]: can we define a type for that?

File: src/communication/WebSocket.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+    send(message : string) {
+        this.ws.send(message);
+    }
+
+    close() {
+        this.ws.close();
+    }
+
+    //event  == onopen onclose onerror onmessage
```
Comment (maurelio1234) [inline src/communication/WebSocket.ts]: ???

File: src/communication/xcConnection.ts [hunk 8]
```diff
@@ -2,45 +2,62 @@ import SessionFactory from "communication/xcSession";
...
-            let parser = new Parser();
-            let configuration = new Configuration(parser);
-            configuration.init(xcApi);
-            session.configuration = configuration;
-            session.replyPublisher.configuration = configuration;
-            sessionListener(null, session);
-        });
+class Connection {
+
+    private apis : any;
```
Comment (maurelio1234) [inline src/communication/xcConnection.ts]: could we define a type for that?

File: src/communication/xcConnection.ts [hunk 9]
```diff
@@ -2,45 +2,62 @@ import SessionFactory from "communication/xcSession";
...
+        this.init(xcApiFileName, serverUrl, null, sessionListener);
+    };
+
+    createAuthenticatedSession(xcApiFileName, serverUrl, sessionData, sessionListener) {
+        this.init(xcApiFileName, serverUrl, sessionData, sessionListener);
+    };
+
+    private init(xcApiFileName, serverUrl, sessionData, sessionListener) {
+        let session = SessionFactory.create(serverUrl, null, sessionData);
+        let thisObject = this;
```
Comment (maurelio1234) [inline src/communication/xcConnection.ts]: thisConnection?

File: src/communication/xcSession.ts [hunk 10]
```diff
@@ -1,141 +1,167 @@
...
+import IWebSocket from "communication/IWebSocket";
+import WebSocket from "communication/WebSocket";
+
+class Session {
+
+    public serverUrl : string;
+    public webSocket : IWebSocket;
+    public configuration : Configuration;
+    public sessionData : string;
+    public guid : any;
```
Comment (maurelio1234) [inline src/communication/xcSession.ts]: can we define a type for that?

File: src/communication/xcSession.ts [hunk 11]
```diff
@@ -1,141 +1,167 @@
...
+    public publishers : Array<Publisher>;
+    public subscribers : Array<Subscriber>;
+    public privateTopics : Array<String>;
+
+    constructor(serverUrl : string, webSocket : IWebSocket, configuration : Configuration, sessionData: string) {
+        this.serverUrl = serverUrl;
+        this.webSocket = webSocket;
+        this.configuration = configuration;
+        this.guid = new Guid();
+        this.privateTopic = this
```
Comment (maurelio1234) [inline src/communication/xcSession.ts]: one line?

File: src/communication/xcSession.ts [hunk 12]
```diff
@@ -1,141 +1,167 @@
...
-};
-
-
-Session.prototype.createSubscriber = function () {
-    let subscriber = new Subscriber(this.webSocket, this.configuration, this.replyPublisher, this.guid, this.privateTopics);
-    this.subscribers.push(subscriber);
-    return subscriber;
-};
+    init(sessionListener, getXcApiRequest, xcApiFileName) {
+        let thisObject = this;
```
Comment (maurelio1234) [inline src/communication/xcSession.ts]: this session?

File: src/communication/xcSession.ts [hunk 13]
```diff
@@ -1,141 +1,167 @@
...
+
+        this.webSocket.setEventListener('onopen', function (e) {
```
Comment (maurelio1234) [inline src/communication/xcSession.ts]: can we define a type for the e's ? in the other functions too.

File: src/communication/xcWebSocketPublisher.ts [hunk 14]
```diff
@@ -1,100 +1,131 @@
...
-let Publisher = function (webSocket, configuration, privateTopic, sessionData) {
-    this.webSocket = webSocket;
-    this.configuration = configuration;
-    this.privateTopic = privateTopic;
-    this.sessionData = sessionData;
-};
-
+class Publisher {
+
+    public webSocket: any;
```
Comment (maurelio1234) [inline src/communication/xcWebSocketPublisher.ts]: could we define types for those attributes?

File: src/communication/xcWebSocketPublisher.ts [hunk 15]
```diff
@@ -1,100 +1,131 @@
...
+    public configuration: any;
+    public privateTopic: any;
+    public sessionData: any;
+
+    constructor(webSocket, configuration, privateTopic, sessionData) {
```
Comment (maurelio1234) [inline src/communication/xcWebSocketPublisher.ts]: types.

Comment (maurelio1234) [inline src/communication/xcWebSocketPublisher.ts]: the same for the other methods...

File: src/communication/xcWebSocketPublisher.ts [hunk 16]
```diff
@@ -1,100 +1,131 @@
...
-    let codes = this.configuration.getCodes(componentName, stateMachineName);
-    let headerConfig = this.getHeaderConfig(codes.componentCode, codes.stateMachineCode, messageType, visibilityPrivate, specifiedPrivateTopic);
-    let event = {
-        "Header": headerConfig.header,
-        "JsonMessage": JSON.stringify(jsonMessage)
-    };
-    return {
-        event: event,
-        routingKey: headerConfig.routingKey
+    getEventToSend(componentName, stateMachineName, messageType, jsonMessage, visibilityPrivate, specifiedPrivateTopic) {
```
Comment (maurelio1234) [inline src/communication/xcWebSocketPublisher.ts]: can we define a return type?

File: src/communication/xcWebSocketPublisher.ts [hunk 17]
```diff
@@ -1,100 +1,131 @@
...
     };
-};
 
-
-Publisher.prototype.canPublish = function (componentName, stateMachineName, messageType) {
-    if (this.configuration.codesExist(componentName, stateMachineName)) {
-        let codes = this.configuration.getCodes(componentName, stateMachineName);
-        if (this.configuration.publisherExist(codes.componentCode, codes.stateMachineCode, messageType)) {
-            return true;
+    canPublish(componentName, stateMachineName, messageType) {
```
Comment (maurelio1234) [inline src/communication/xcWebSocketPublisher.ts]: return type?

File: src/communication/xcWebSocketPublisher.ts [hunk 18]
```diff
@@ -1,100 +1,131 @@
...
-let convertToWebsocketInputFormat = function (data) {
-    let input = data.routingKey + " " + data.event.Header.ComponentCode.Fields[0] +
-        " " + JSON.stringify(data.event);
-    return input;
-};
+    getHeaderConfig(componentCode, stateMachineCode, messageType, visibilityPrivate, specifiedPrivateTopic) {
+        let publisher = this
+            .configuration
+            .getPublisherDetails(componentCode, stateMachineCode, messageType);
+        let thisObject = this;
```
Comment (maurelio1234) [inline src/communication/xcWebSocketPublisher.ts]: this... Publisher?

File: src/configuration/xcConfiguration.ts [hunk 19]
```diff
@@ -1,68 +1,81 @@
...
-    return this.parser.getSnapshotTopic(componentName);
-};
-
-
-Configuration.prototype.getStateName = function (componentCode, stateMachineCode, stateCode) {
-    return this.parser.getStateName(componentCode, stateMachineCode, stateCode);
-};
+class Configuration {
+
+    private parser : any;
```
Comment (maurelio1234) [inline src/configuration/xcConfiguration.ts]: type?

File: src/configuration/xcConfiguration.ts [hunk 20]
```diff
@@ -1,68 +1,81 @@
...
+        this.parser = parser;
+    }
+
+    init(xml) {
+        this
+            .parser
+            .parse(xml, this.tags);
+    };
+
+    getCodes(componentName, stateMachineName) {
```
Comment (maurelio1234) [inline src/configuration/xcConfiguration.ts]: the same thing as before... parameter types and return types.

File: src/communication/xcSession.ts [hunk 21]
```diff
@@ -1,141 +1,167 @@
...
+        });
+
+        this.webSocket.setEventListener('onerror', function (e) {
+            let messageError = "Error on " + thisObject.serverUrl + ".";
+            console.error(messageError);
+            console.error(e);
+            sessionListener(messageError, null);
+        });
+
+        this.webSocket.setEventListener('onclose', function (e) {
```
Comment (maurelio1234) [inline src/communication/xcSession.ts]: do we control these names? onClose seems much better...

File: src/communication/IWebSocket.ts [hunk 22]
```diff
@@ -0,0 +1,10 @@
+interface IWebSocket
```
Comment (nicolaserny) [inline src/communication/IWebSocket.ts]: A standard WebSocket is defined in typescript by default. Just use it

File: src/javascriptHelper.ts [hunk 23]
```diff
@@ -1,12 +1,20 @@
```
Comment (nicolaserny) [inline src/javascriptHelper.ts]: You didn't test your code. Merge incomplete it cannot work !

File: src/javascriptHelper.ts [hunk 24]
```diff
@@ -1,20 +1,12 @@
 
 interface JavascriptHelper {
-<<<<<<< HEAD
     WebSocket: WebSocket;
     DOMParser: any;
```
Comment (nicolaserny) [inline src/javascriptHelper.ts]: Wrong merge - DOMParser is not used anymore. You can remove it

File: src/communication/xcConnection.ts [hunk 25]
```diff
@@ -1,48 +1,54 @@
...
+
+    private apis : {
+        [xcApiFileName : string]: ApiConfiguration
+    };
+
+    constructor() {
+        this.apis = {};
+    }
+
+    getXcApiList(serverUrl : string, getXcApiListListener : (apis : Array < Object >) => void) : void {
```
Comment (nicolaserny) [inline src/communication/xcConnection.ts]: Array<Object> perhaps you can improve the type. Array<Api> for example...

File: src/communication/xcConnection.ts [hunk 26]
```diff
@@ -1,48 +1,54 @@
...
+        let session = SessionFactory(serverUrl, null, null);
+        session.webSocket.onopen = function (e : Event) {
```
Comment (nicolaserny) [inline src/communication/xcConnection.ts]: if you don't use the event e, just use the wild card notation: function (_: event)

File: src/communication/xcConnection.ts [hunk 27]
```diff
@@ -1,48 +1,54 @@
...
+            session
+                .privateSubscriber
+                .getXcApiList(function (apis) {
+                    getXcApiListListener(apis);
+                    session.close();
+                });
+        };
+    };
+
+    createSession(xcApiFileName : string, serverUrl : string, sessionListener : (error : any, session : Session) => void) : void {
```
Comment (nicolaserny) [inline src/communication/xcConnection.ts]: Improve the error type.
Try to avoid any

File: src/communication/xcConnection.ts [hunk 28]
```diff
@@ -1,48 +1,54 @@
...
+        this.init(xcApiFileName, serverUrl, null, sessionListener);
+    };
+
+    createAuthenticatedSession(xcApiFileName : string, serverUrl : string, sessionData : string, sessionListener : (error : any, session : Session) => void) : void {
```
Comment (nicolaserny) [inline src/communication/xcConnection.ts]: error type

File: src/communication/xcConnection.ts [hunk 29]
```diff
@@ -1,48 +1,54 @@
...
+        this.init(xcApiFileName, serverUrl, sessionData, sessionListener);
     };
-    session.init(sessionListener, getXcApiRequest, xcApiFileName);
-};
 
+    private init(xcApiFileName : string, serverUrl : string, sessionData : string, sessionListener : (error : any, session : Session) => void) : void {
```
Comment (nicolaserny) [inline src/communication/xcConnection.ts]: error type

Comment (nicolaserny) [inline src/communication/xcConnection.ts]: Optional: the init method needs refactoring. But we'll do it in another pull request

File: src/communication/xcConnection.ts [hunk 30]
```diff
@@ -1,48 +1,54 @@
...
-                session.replyPublisher.configuration = configuration;
-                sessionListener(null, session);
-            })
-            .catch(e => sessionListener(e, null));
-        });
+import {SessionFactory, Session} from "communication/xcSession";
+import {ApiConfiguration} from "configuration/apiConfiguration";
+import {DefaultApiConfigurationParser} from "configuration/apiConfigurationParser";
+
+class Connection {
```
Comment (nicolaserny) [inline src/communication/xcConnection.ts]: Define an interface Connection and a default implementation

File: src/communication/xcSession.ts [hunk 31]
```diff
@@ -1,145 +1,161 @@
...
-};
-
-
-Session.prototype.setPrivateTopic = function (privateTopic) {
-    this.addPrivateTopic(privateTopic);
-    this.removePrivateTopic(this.privateTopic);
-    this.privateTopic = privateTopic;
-    for (let i = 0; i < this.publishers.length; i++) {
-        this.publishers[i].privateTopic = privateTopic;
+export class Session {
```
Comment (nicolaserny) [inline src/communication/xcSession.ts]: likewise: define an interface Session and a default implementation

File: src/communication/xcSession.ts [hunk 32]
```diff
@@ -1,145 +1,161 @@
...
-
-    this.webSocket.onopen = function (e) {
-        thisObject.privateSubscriber.sendSubscribeRequestToTopic(thisObject.privateTopic, xcWebSocketBridgeConfiguration.kinds.Private);
-        getXcApiRequest(xcApiFileName, sessionListener);
-        console.log("connection started on " + thisObject.serverUrl + ".");
+    setPrivateTopic(privateTopic : string) : void {
+        this.addPrivateTopic(privateTopic);
+        this.removePrivateTopic(this.privateTopic);
+        this.privateTopic = privateTopic;
+        for (let i = 0; i < this.publishers.length; i++) {
```
Comment (nicolaserny) [inline src/communication/xcSession.ts]: optional: Best practice use forEach
https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines

File: src/communication/xcSession.ts [hunk 33]
```diff
@@ -1,145 +1,161 @@
...
+            this.publishers[i].privateTopic = privateTopic;
+        }
+        for (let j = 0; j < this.subscribers.length; j++) {
```
Comment (nicolaserny) [inline src/communication/xcSession.ts]: likewise with forEach

File: src/communication/xcSession.ts [hunk 34]
```diff
@@ -1,145 +1,161 @@
...
-Array.prototype.removeElement = function (e, msg) {
-    let index = this.indexOf(e);
-    if (index > -1) {
-        return this.splice(index, 1);
-    } else {
-        throw new Error(msg);
-    }
-};
 
+    init(sessionListener: (error : any, session : Session) => void, getXcApiRequest : (xcApiFileName : string, sessionListener : (error : any, session : Session) => void) => void, xcApiFileName : string) : void {
```
Comment (nicolaserny) [inline src/communication/xcSession.ts]: Likewise with error type

File: src/communication/xcSession.ts [hunk 35]
```diff
@@ -1,145 +1,161 @@
...
+        this
+            .subscribers
+            .push(subscriber);
+        return subscriber;
+    };
 
-Session.prototype.disposeSubscriber = function (subscriber) {
-    this.subscribers.removeElement(subscriber, "Subscriber not found");
-};
+    private removeElement(array : Array<Object>, e : Object, msg : string) : void{
```
Comment (nicolaserny) [inline src/communication/xcSession.ts]: use typescript generics [https://www.typescriptlang.org/docs/handbook/generics.html](url)
Ex: array: Array<T>, e: T

File: src/communication/xcWebSocketPublisher.ts [hunk 36]
```diff
@@ -1,102 +1,132 @@
...
 import { ApiConfiguration } from "configuration/apiConfiguration";
 
-let Publisher = function (webSocket, configuration: ApiConfiguration, privateTopic, sessionData) {
-    this.webSocket = webSocket;
-    this.configuration = configuration;
-    this.privateTopic = privateTopic;
-    this.sessionData = sessionData;
-};
-
+class Publisher {
```
Comment (nicolaserny) [inline src/communication/xcWebSocketPublisher.ts]: Create Publisher interface

File: src/communication/xcWebSocketPublisher.ts [hunk 37]
```diff
@@ -1,102 +1,132 @@
...
+        return {
+            event: event,
+            routingKey: headerConfig.routingKey
+        };
     };
-};
 
-
-Publisher.prototype.canPublish = function (componentName, stateMachineName, messageType) {
+    canPublish(componentName : string, stateMachineName : string, messageType : string) {
```
Comment (nicolaserny) [inline src/communication/xcWebSocketPublisher.ts]: Can messageType be an enum ?

File: src/communication/xcWebSocketPublisher.ts [hunk 38]
```diff
@@ -1,102 +1,132 @@
...
+            .webSocket
+            .send(webSocketInputFormat);
+    };
 
-let convertToWebsocketInputFormat = function (data) {
-    let input = data.routingKey + " " + data.event.Header.ComponentCode.Fields[0] +
-        " " + JSON.stringify(data.event);
-    return input;
-};
+    getHeaderConfig(componentCode : number, stateMachineCode : number, messageType : string, visibilityPrivate : boolean, specifiedPrivateTopic  : string) {
```
Comment (nicolaserny) [inline src/communication/xcWebSocketPublisher.ts]: Define an explicit return type

Comment (nicolaserny) [inline src/communication/xcWebSocketPublisher.ts]: Could be interesting to define an Header type too

**Comment by nicolaserny** [comment]
test this please

File: src/javascriptHelper.ts (hunk 24 above)
Comment (mbarekh) [inline src/javascriptHelper.ts]: done

File: src/communication/xcConnection.ts (hunk 30 above)
Comment (mbarekh) [inline src/communication/xcConnection.ts]: done

File: src/communication/xcConnection.ts (hunk 25 above)
Comment (mbarekh) [inline src/communication/xcConnection.ts]: done

File: src/communication/xcConnection.ts (hunk 26 above)
Comment (mbarekh) [inline src/communication/xcConnection.ts]: done

File: src/communication/xcConnection.ts (hunk 27 above)
Comment (mbarekh) [inline src/communication/xcConnection.ts]: done

File: src/communication/xcConnection.ts (hunk 28 above)
Comment (mbarekh) [inline src/communication/xcConnection.ts]: done

File: src/communication/xcConnection.ts (hunk 29 above)
Comment (mbarekh) [inline src/communication/xcConnection.ts]: error type done

File: src/communication/xcSession.ts (hunk 31 above)
Comment (mbarekh) [inline src/communication/xcSession.ts]: done

File: src/communication/xcSession.ts (hunk 33 above)
Comment (mbarekh) [inline src/communication/xcSession.ts]: done

File: src/communication/xcSession.ts (hunk 34 above)
Comment (mbarekh) [inline src/communication/xcSession.ts]: done

File: src/communication/xcSession.ts (hunk 35 above)
Comment (mbarekh) [inline src/communication/xcSession.ts]: done

File: src/communication/xcSession.ts [hunk 39]
```diff
@@ -6,140 +6,157 @@ import xcWebSocketBridgeConfiguration from "configuration/xcWebSocketBridgeConfi
...
+
+    private serverUrl: string;
+    private sessionData: string;
+    private guid: Guid;
+    private privateTopic: string;
+    private publishers: Array<Publisher>;
+    private subscribers: Array<Subscriber>;
+    private privateTopics: Array<String>;
+
+    public privateSubscriber: Subscriber;
```
Comment (nicolaserny) [inline src/communication/xcSession.ts]: members are public by default. No need to use public

File: src/communication/xcSession.ts [hunk 40]
```diff
@@ -6,140 +6,157 @@ import xcWebSocketBridgeConfiguration from "configuration/xcWebSocketBridgeConfi
...
-    this.privateTopics.push(privateTopic);
-    for (let i = 0; i < this.subscribers.length; i++) {
-        this.subscribers[i].privateTopics = this.privateTopics;
-    }
-};
+    setPrivateTopic(privateTopic: string): void {
+        this.addPrivateTopic(privateTopic);
+        this.removePrivateTopic(this.privateTopic);
+        this.privateTopic = privateTopic;
+        this.publishers.forEach(function (publisher) {
```
Comment (nicolaserny) [inline src/communication/xcSession.ts]: Style: Use arrow functions over anonymous function expressions
[https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines](url)

File: src/communication/xcSession.ts [hunk 41]
```diff
@@ -6,140 +6,157 @@ import xcWebSocketBridgeConfiguration from "configuration/xcWebSocketBridgeConfi
...
+            publisher.privateTopic = privateTopic;
+        });
+        this.subscribers.forEach(function (subscriber) {
```
Comment (nicolaserny) [inline src/communication/xcSession.ts]: use arrow functions

File: src/communication/xcSession.ts [hunk 42]
```diff
@@ -6,140 +6,157 @@ import xcWebSocketBridgeConfiguration from "configuration/xcWebSocketBridgeConfi
...
-    this.privateTopics.removeElement(privateTopic, "private topic not found");
-    for (let i = 0; i < this.subscribers.length; i++) {
-        this.subscribers[i].privateTopics = this.privateTopics;
-    }
-};
+    addPrivateTopic(privateTopic: string): void {
+        let kindPrivate = xcWebSocketBridgeConfiguration.kinds.Private;
+        this.privateSubscriber.sendSubscribeRequestToTopic(privateTopic, kindPrivate);
+        this.privateTopics.push(privateTopic);
+        this.subscribers.forEach(function (subscriber) {
```
Comment (nicolaserny) [inline src/communication/xcSession.ts]: use arrow function

File: src/communication/xcSession.ts [hunk 43]
```diff
@@ -6,140 +6,157 @@ import xcWebSocketBridgeConfiguration from "configuration/xcWebSocketBridgeConfi
...
-Session.prototype.removePrivateTopic = function (privateTopic) {
-    let kindPrivate = xcWebSocketBridgeConfiguration.kinds.Private;
-    this.privateSubscriber.sendUnsubscribeRequestToTopic(privateTopic, kindPrivate);
-    this.privateTopics.removeElement(privateTopic, "private topic not found");
-    for (let i = 0; i < this.subscribers.length; i++) {
-        this.subscribers[i].privateTopics = this.privateTopics;
-    }
-};
+    addPrivateTopic(privateTopic: string): void {
+        let kindPrivate = xcWebSocketBridgeConfiguration.kinds.Private;
```
Comment (nicolaserny) [inline src/communication/xcSession.ts]: use const instead of let

File: src/communication/xcSession.ts [hunk 44]
```diff
@@ -6,140 +6,157 @@ import xcWebSocketBridgeConfiguration from "configuration/xcWebSocketBridgeConfi
...
+        this.privateSubscriber.sendSubscribeRequestToTopic(privateTopic, kindPrivate);
+        this.privateTopics.push(privateTopic);
+        this.subscribers.forEach(function (subscriber) {
+            subscriber.privateTopics = this.privateTopics;
+        }, this);
+    };
 
+    removePrivateTopic(privateTopic: string): void {
+        let kindPrivate = xcWebSocketBridgeConfiguration.kinds.Private;
```
Comment (nicolaserny) [inline src/communication/xcSession.ts]: use const...

File: src/communication/xcSession.ts [hunk 45]
```diff
@@ -6,140 +6,157 @@ import xcWebSocketBridgeConfiguration from "configuration/xcWebSocketBridgeConfi
...
+        let thisSession = this;
+
+        this.webSocket.onopen = function (e: Event) {
+            thisSession.privateSubscriber.sendSubscribeRequestToTopic(thisSession.privateTopic, xcWebSocketBridgeConfiguration.kinds.Private);
+            openListener(e);
+            console.log("connection started on " + thisSession.serverUrl + ".");
+        };
+
+        this.webSocket.onerror = function (e: Event) {
+            let messageError = "Error on " + thisSession.serverUrl + ".";
```
Comment (nicolaserny) [inline src/communication/xcSession.ts]: use const

File: src/communication/xcSession.ts [hunk 46]
```diff
@@ -6,140 +6,157 @@ import xcWebSocketBridgeConfiguration from "configuration/xcWebSocketBridgeConfi
...
-    for (let j = 0; j < this.subscribers.length; j++) {
-        this.disposeSubscriber(this.subscribers[j]);
-    }
-};
+    disposeSubscriber(subscriber: Subscriber): void {
+        this.removeElement(this.subscribers, subscriber);
+    };
 
+    dispose(): void {
+        this.publishers.forEach(function (publisher) {
```
Comment (nicolaserny) [inline src/communication/xcSession.ts]: use arrow function

File: src/communication/xcSession.ts [hunk 47]
```diff
@@ -6,140 +6,157 @@ import xcWebSocketBridgeConfiguration from "configuration/xcWebSocketBridgeConfi
...
+            this.disposePublisher(publisher);
+        }, this);
+        this.subscribers.forEach(function (subscriber) {
```
Comment (nicolaserny) [inline src/communication/xcSession.ts]: use arrow function

File: src/communication/xcSession.ts [hunk 48]
```diff
@@ -6,140 +6,157 @@ import xcWebSocketBridgeConfiguration from "configuration/xcWebSocketBridgeConfi
...
+        }, this);
+    };
 
-Session.prototype.close = function () {
-    this.privateSubscriber.sendUnsubscribeRequestToTopic(this.privateTopic, xcWebSocketBridgeConfiguration.kinds.Private);
-    this.dispose();
-    this.webSocket.close();
-};
+    close(): void {
+        this.privateTopics.forEach(function (privateTopic) {
```
Comment (nicolaserny) [inline src/communication/xcSession.ts]: use arrow function

File: src/communication/xcSession.ts [hunk 49]
```diff
@@ -6,140 +6,157 @@ import xcWebSocketBridgeConfiguration from "configuration/xcWebSocketBridgeConfi
...
+        this.webSocket.close();
+    };
 
+}
 
-let returnObject = {
-    create: SessionFactory,
-    Session: Session
+export let SessionFactory = function (serverUrl: string, configuration: ApiConfiguration, sessionData: string): Session {
+    let webSocket = new WebSocket(serverUrl);
```
Comment (nicolaserny) [inline src/communication/xcSession.ts]: use const

File: src/communication/xcSession.ts [hunk 50]
```diff
@@ -6,140 +6,157 @@ import xcWebSocketBridgeConfiguration from "configuration/xcWebSocketBridgeConfi
...
+    let session = new DefaultSession(serverUrl, webSocket, configuration, sessionData);
```
Comment (nicolaserny) [inline src/communication/xcSession.ts]: use const

File: src/communication/xcWebSocketPublisher.ts [hunk 51]
```diff
@@ -1,102 +1,132 @@
...
+            .webSocket
+            .send(webSocketInputFormat);
+    };
 
-let convertToWebsocketInputFormat = function (data) {
-    let input = data.routingKey + " " + data.event.Header.ComponentCode.Fields[0] +
-        " " + JSON.stringify(data.event);
-    return input;
-};
+    getHeaderConfig(componentCode: number, stateMachineCode: number, messageType: string, visibilityPrivate: boolean, specifiedPrivateTopic: string) {
```
Comment (nicolaserny) [inline src/communication/xcWebSocketPublisher.ts]: define explicit return type

File: src/configuration/xcWebSocketBridgeConfiguration.ts [hunk 52]
```diff
@@ -6,6 +6,8 @@ let kinds = {
 };
 
 let commands = {
```
Comment (nicolaserny) [inline src/configuration/xcWebSocketBridgeConfiguration.ts]: looks like an enum

File: src/javascriptHelper.ts [hunk 53]
```diff
@@ -1,11 +1,12 @@
...
 interface JavascriptHelper {
-    WebSocket: any;
+    WebSocket: WebSocket;
+    DOMParser: any;
     atob: any;
 }
 
 const javascriptHelper = () : JavascriptHelper => {
-    return { WebSocket: window["WebSocket"], atob: window.atob };
+    return { WebSocket: <WebSocket>window["WebSocket"], DOMParser: window["DOMParser"], atob: window.atob };
```
Comment (nicolaserny) [inline src/javascriptHelper.ts]: remove DOMParser

File: test/index.html [hunk 54]
```diff
@@ -0,0 +1,11 @@
+<!DOCTYPE html>
```
Comment (nicolaserny) [inline test/index.html]: Avoid to push theses tests on the repository

File: src/communication/xcSession.ts [hunk 55]
```diff
@@ -6,140 +6,157 @@ import xcWebSocketBridgeConfiguration from "configuration/xcWebSocketBridgeConfi
...
+        this.subscribers.push(subscriber);
+        return subscriber;
     };
 
-    this.webSocket.onclose = function (e) {
-        console.log("connection on " + thisObject.serverUrl + " closed.");
-        console.log(e);
-        if (!e.wasClean) {
-            throw new Error("unexpected connection close with code " + e.code);
+    private removeElement<T>(array: Array<T>, e: T): void {
```
Comment (nicolaserny) [inline src/communication/xcSession.ts]: Create an extension method for Array.
This method is already defined in definition.d.ts
